File: go/structural/decorator/ChaiDecorator.go
package decorator

import (
	"strings"
)


type ChaiDecorator struct {
    teaToMakeChai Tea
    chaiIngredients []string
}

func NewChaiDecorator( teaToMakeChai Tea ) *ChaiDecorator {
	obj := new(ChaiDecorator)
	obj.teaToMakeChai = teaToMakeChai
	obj.chaiIngredients = make([]string,0);
	obj.chaiIngredients = append(obj.chaiIngredients,"bay leaf")
	obj.chaiIngredients = append(obj.chaiIngredients,"cinnamon stick")
	obj.chaiIngredients = append(obj.chaiIngredients,"ginger")
	obj.chaiIngredients = append(obj.chaiIngredients,"honey")
	obj.chaiIngredients = append(obj.chaiIngredients,"soy milk")
	obj.chaiIngredients = append(obj.chaiIngredients,"vanilla bean")
	return obj
}

func (this *ChaiDecorator) SteepTea() string {
	return this.steepChai()
}

func (this *ChaiDecorator) steepChai() string {
	return this.teaToMakeChai.SteepTea() + this.steepChaiIngredients()
}    
    
func (this *ChaiDecorator) steepChaiIngredients() string {
	retval := make([]string,0)
	for _,v:= range this.chaiIngredients {
		retval = append( retval, v + " is steeping" )
	}
	return strings.Join(retval,", ")
}

File: scala/src/main/scala/Structural/Decorator/ChaiDecorator.scala
package Structural.Decorator;


//ChaiDecorator - the decorator

import java.util.*;
import org.apache.commons.lang3.StringUtils;

class ChaiDecorator extends Tea {
    private Tea teaToMakeChai;
    private ArrayList chaiIngredients = new ArrayList();
    
    ChaiDecorator(Tea teaToMakeChai) {
        this.addTea(teaToMakeChai);
        chaiIngredients.add("bay leaf");
        chaiIngredients.add("cinnamon stick");
        chaiIngredients.add("ginger");
        chaiIngredients.add("honey");
        chaiIngredients.add("soy milk");
        chaiIngredients.add("vanilla bean");
    }

    private void addTea(Tea teaToMakeChaiIn) {
        this.teaToMakeChai = teaToMakeChaiIn;
    }
    
    String steepTea() {
        return this.steepChai();
    }

    String steepChai() {
        teaToMakeChai.steepTea();
        return this.steepChaiIngredients();
    }    
    
    String steepChaiIngredients() {
        ListIterator listIterator = chaiIngredients.listIterator();
		 List<String> retval = new ArrayList<String>();
        while (listIterator.hasNext()) {
				retval.add(listIterator.next() + " is steeping");
        }
        return StringUtils.join(retval,", ");
    }      
}
